device-scanner/mount-emitter/src/lib.rs

```diff
@@ -8,67 +8,94 @@
 //! into a `MountCommand` variant and send the serialized result to `device-scanner`.
 //!
 
-use std::convert::AsRef;
-
 use device_types::{
     mount::{FsType, MountCommand, MountOpts, MountPoint},
     Command, DevicePath,
 };
-use futures::{future, Future};
-use std::{
-    collections::HashMap, io::BufRead, os::unix::net::UnixStream as NetUnixStream, process::exit,
-    str,
-};
+use futures::Future;
+use std::{collections::HashMap, convert::AsRef, os::unix::net::UnixStream as NetUnixStream, str};
 use tokio::{
-    io::{AsyncRead, AsyncWrite},
+    io::{self, AsyncBufRead, AsyncBufReadExt, AsyncWrite, AsyncWriteExt},
     net::UnixStream,
-    reactor::Handle,
 };
 
-fn line_to_command(x: &[u8]) -> MountCommand {
-    let mut x: IntermediateMap = line_to_hashmap(&x);
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+    #[error(transparent)]
+    SerdeJson(#[from] serde_json::Error),
+    #[error(transparent)]
+    Utf8Error(#[from] std::str::Utf8Error),
+    #[error("missing {0}")]
+    Missing(&'static str),
+    #[error("unexpected {0}")]
+    Unexpected(String),
+}
+
+pub fn line_to_command(x: &[u8]) -> Result<MountCommand, Error> {
+    let mut x: IntermediateMap = line_to_hashmap(&x)?;
 
     let (target, source, fstype, mount_opts, old_opts, old_target) = (
-        MountPoint(x.remove("TARGET").expect("missing TARGET").into()),
-        DevicePath(x.remove("SOURCE").expect("missing SOURCE").into()),
-        FsType(x.remove("FSTYPE").expect("missing FSTYPE")),
-        MountOpts(x.remove("OPTIONS").expect("missing OPTIONS")),
+        MountPoint(
+            x.remove("TARGET")
+                .ok_or_else(|| Error::Missing("TARGET"))?
+                .into(),
+        ),
+        DevicePath(
+            x.remove("SOURCE")
+                .ok_or_else(|| Error::Missing("SOURCE"))?
```

Comment:
Doesn't need to be `_else`?

Reply:
Wasn't sure if clippy would complain about the allocation